test/testsuite/TestSuite.jl

```diff
@@ -0,0 +1,69 @@
+# Based on the design of GPUArrays.jl
+
+"""
+    TestSuite
+
+Suite of tests that may be used for all packages inheriting from MatrixAlgebraKit.
+
+"""
+module TestSuite
+
+using Test, TestExtras
+using MatrixAlgebraKit
+using MatrixAlgebraKit: diagview
+using LinearAlgebra: Diagonal, norm, istriu, istril
+using Random, StableRNGs
+using AMDGPU, CUDA
+
+const tests = Dict()
+
+macro testsuite(name, ex)
+    safe_name = lowercase(replace(replace(name, " " => "_"), "/" => "_"))
+    fn = Symbol("test_", safe_name)
+    return quote
+        $(esc(fn))(AT; eltypes = supported_eltypes(AT, $(esc(fn)))) = $(esc(ex))(AT, eltypes)
+        @assert !haskey(tests, $name) "testsuite already exists"
+        tests[$name] = $fn
+    end
+end
+
+testargs_summary(args...) = string(args)
+
+const rng = StableRNG(123)
+seed_rng!(seed) = Random.seed!(rng, seed)
+
+instantiate_matrix(::Type{T}, size) where {T <: Number} = randn(rng, T, size)
+instantiate_matrix(::Type{AT}, size) where {AT <: Array} = randn(rng, eltype(AT), size)
+instantiate_matrix(::Type{AT}, size) where {AT <: CuArray} = CuArray(randn(rng, eltype(AT), size))
+instantiate_matrix(::Type{AT}, size) where {AT <: ROCArray} = ROCArray(randn(rng, eltype(AT), size))
+instantiate_matrix(::Type{AT}, size) where {AT <: Diagonal} = Diagonal(randn(rng, eltype(AT), size))
+instantiate_matrix(::Type{AT}, size) where {T, AT <: Diagonal{T, <:CuVector}} = Diagonal(CuArray(randn(rng, eltype(AT), size)))
+instantiate_matrix(::Type{AT}, size) where {T, AT <: Diagonal{T, <:ROCVector}} = Diagonal(ROCArray(randn(rng, eltype(AT), size)))
+
+precision(::Type{T}) where {T <: Number} = sqrt(eps(real(T)))
+precision(::Type{T}) where {T} = precision(eltype(T))
+
+function has_positive_diagonal(A)
+    T = eltype(A)
+    return if T <: Real
+        all(≥(zero(T)), diagview(A))
+    else
+        all(≥(zero(real(T))), real(diagview(A))) &&
+            all(≈(zero(real(T))), imag(diagview(A)))
+    end
```

Comment:
```suggestion
    all(x -> x ≈ abs(x), diagview(A))
```

Reply:
We can do this, but it does change the tests, `-eps()` would have this property but it wouldn't have been properly "positive". `sign(x) == 1` might work but I'm not sure how accurate that is

Reply:
Well indeed, I also think we currently require strict positivity, but I was confused by `all(≈(zero(real(T))), imag(diagview(A)))`. Is this equivalent to `all(iszero, imag(diagview(A)))`. If you want the latter, then clearly we want `x == abs(x)`.

Reply:
or simply `all(isposdef, diagview(A))` (but this requires LinearAlgebra.isposdef)